### PR TITLE
improve(across-v2-umip): Add more details and definitions, and additional rules implemented in Dataworker

### PR DESCRIPTION
**Pull request by nicholaspai**
The UMIP was proposed and merged prior to the [Dataworker](https://github.com/across-protocol/relayer-v2/tree/master/src/dataworker) being implemented, which is the piece of code that is supposed to construct root bundles following the rules of the UMIP. In the process of writing the Dataworker code, we realized several edge cases in the UMIP. This PR adds in rules to handle those cases. Additionally, this PR adds a lot of details and resources to better clarify the UMIP. 
﻿

### CODE REVIEW

**Comment by nicholaspai** [comment]
Note @mrice32 I have not touched some of the lower sections on constructing the actual roots, as I think most of the specific rules have not changed. But some of the language in the above sections have. I did fix one issue we found where I think the running balances don't explicitly instruct to add Slow Fill payments

File: UMIPs/umip-157.md [hunk 1]
```diff
@@ -12,44 +12,59 @@
...
 The DVM should support the ACROSS-V2 price identifier.
 
-## Across V2 Architecture
+# Across V2 Architecture
 
-The basic architecture of Across V2 is a single LP pool sitting on Ethereum mainnet, while many "spokes" are used on
-various chains to facilitate user "deposits" where a cross-chain transfer is initiated and "withdrawals" where a user
-is relayed the funds they transferred from another chain.
+The basic architecture of Across V2 is a single LP ("Liquidity Provider") pool sitting on Ethereum mainnet connected to many "spoke pools" deployed on
+various chains to facilitate user "deposits". A deposit is a cross-chain transfer request from an "origin" chain to a different "destination" chain, which is fulfilled when a "relayer" sends the depositor their desired transfer amount (less fees) on their desired destination chain.
```
Comment (mrice32) [inline UMIPs/umip-157.md:18]: +1

File: UMIPs/umip-157.md [hunk 2]
```diff
@@ -12,44 +12,59 @@
...
 
-If there is no relayer who can provide all the capital, a slow relay is performed where the funds are sent from the
-LP pool via the bundling process and anyone can use those funds to provide the remaining required capital to the user.
 
-The LP pool verification process works by having a proposer propose a bundle, which is effectively a group of merkle
-roots for a range of block numbers they specify in their proposal. This UMIP verifies that bundling process.
+If there is no relayer who can provide all the capital for a given deposit request, a "slow relay" (or "slow fill") is performed where the funds are sent from the
+LP pool to the destination spoke to fulfill the deposit. These slow fill requests are also included in the aformentioned Bundles.
```
Comment (mrice32) [inline UMIPs/umip-157.md:24]:
```suggestion
LP pool to the destination spoke to fulfill the deposit. These slow fill requests are also included in the aforementioned bundles.
```

File: UMIPs/umip-157.md [hunk 3]
```diff
@@ -12,44 +12,59 @@
...
+
+Bundles are implemented on-chain as [Merkle Roots](https://www.youtube.com/watch?v=JXn4GqyS7Gg) which uniquely identify the set of all repayments and rebalance instructions over a specific time range. Therefore, Across V2 moves capital to repay relayers by continuous Bundles, all validated by the OO.
```
Comment (mrice32) [inline UMIPs/umip-157.md:26]:
```suggestion
Bundles are implemented on-chain as [Merkle Roots](https://www.youtube.com/watch?v=JXn4GqyS7Gg) which uniquely identify the set of all repayments and rebalance instructions over a specific block range. Therefore, Across V2 moves capital to repay relayers and fulfil bridge requests through periodic bundles, all validated by the OO.
```

File: UMIPs/umip-157.md [hunk 4]
```diff
@@ -12,44 +12,59 @@
...
-sort should be on the transactionIndex, and the tertiary sort should be on the logIndex.
+sort should be on the `transactionIndex`, and the tertiary sort should be on the `logIndex`. See the section on [comparing events](#comparing-events-chronologically) for more details.
 
 Note 3: wherever unspecified, sorting should be ascending by default, not descending.
 
-## Ancillary Data Specifications
+# Definitions
+
+## Comparing events chronologically
+Smart contract transactions can emit events that conform to the specifications described in the "Returns" section of these [docs](https://web3js.readthedocs.io/en/v1.2.11/web3-eth-contract.html#contract-events). Specifically, an event is expected to have a `blockNumber`, `transactionIndex` and `logIndex`. To compare events `e1` and `e2` chronologically, we can say
```
Comment (mrice32) [inline UMIPs/umip-157.md:52]:
```suggestion
Smart contract transactions can emit events that conform to the specifications described in the "Returns" section of these [docs](https://web3js.readthedocs.io/en/v1.2.11/web3-eth-contract.html#contract-events). Specifically, an event is expected to have a unique combination of `blockNumber`, `transactionIndex` and `logIndex`. To compare events `e1` and `e2` chronologically, we can say
```

File: UMIPs/umip-157.md [hunk 5]
```diff
@@ -12,44 +12,59 @@
...
+that `e1` is "earlier" than `e2` if `e1.blockNumber < e2.blockNumber` OR if `e1.blockNumber == e2.blockNumber && e1.transactionIndex < e2.transactionIndex` OR if `e1.blockNumber == e2.blockNumber && e1.transactionIndex == e2.transactionIndex && e1.logIndex < e2.logIndex`.
+
+So, "earlier" events have a lower block number, transaction index, or log index, and we should compare event properties in that order.
+
+## Valid bundle proposals
+A valid root bundle submitted by calling `HubPool.proposeRootBundle()` will emit a [`ProposedRootBundle`](https://github.com/across-protocol/contracts-v2/blob/a8ab11fef3d15604c46bba6439291432db17e745/contracts/HubPool.sol#L148-L156) event.
```
Comment (mrice32) [inline UMIPs/umip-157.md:58]:
```suggestion
A root bundle can be proposed by calling `HubPool.proposeRootBundle()`, which will will emit a [`ProposedRootBundle`](https://github.com/across-protocol/contracts-v2/blob/a8ab11fef3d15604c46bba6439291432db17e745/contracts/HubPool.sol#L148-L156) event.
```

**Comment by nicholaspai** [comment]
@mrice32 did you have a list elsewhere of missing features/bugs that i haven't addressed yet in this PR?

**Comment by mrice32** [comment]
@nicholaspai My list:
- Clarify how to sort addresses ✅ 
- Clarify what we mean when we say "matching indices" when matching an L1 token to a running balance ❓
- Update slow relay "excess" algorithm ✅

**Comment by mrice32** [comment]
> > @nicholaspai My list:
> > 
> > * Clarify how to sort addresses ✅
> > * Clarify what we mean when we say "matching indices" when matching an L1 token to a running balance ❓
> > * Update slow relay "excess" algorithm ✅
> 
> nice, added some language for points 2 and 3 (I don't think i addressed 3 despite the check mark) in [this commit](https://github.com/UMAprotocol/UMIPs/pull/550/commits/7dc327243240cad5d7b6afe26b689d49fadcafd4)

Awesome, thanks, reviewing this before EOD!

File: UMIPs/umip-157.md [hunk 6]
```diff
@@ -193,51 +237,42 @@ Similarly, sum the `fillAmount * realizedLPFeePct / 1e18` to get the total LP fe
...
... not have triggered a slow fill payment to be sent to the spoke. First we need to see if the first fill for this current fill triggered a slow fill. Looking through all [valid fills](#valid-bundle-proposals) in history, identify all matching fills for the same `originChainId` and `depositId`. Find the earliest such fill. Determine the block range of the root bundle proposal that contained this fill using [this logic](#determining-block-range-for-root-bundle-proposal) for the `ProposeRootBundle` event with a `bundleEndBlock` for the `FilledRelay.destinationChainId` greater than or equal to the `FilledRelay` block number. Find the last fill in this same bundle block range. The slow fill payment for the bundle should have been `FilledRelay.amount - FilledRelay.totalFilledAmount`, same as [this calculation](#computing-slow-relay-payment-amounts). Since we know that the current `FilledRelay` in this upcoming proposal 100% filled the deposit, we know that the slow fill leaf cannot be executed, so the entire slow fill payment must be sent back to mainnet. Subtract this (the preceiding slow fill payment from a valid root bundle that this fill ultimately completed) from the running balance.
```
Comment (mrice32) [inline UMIPs/umip-157.md:253]: nit: do we specify that this logic should only apply for past bundles?

**Review by mrice32 — APPROVED** [review]
LGTM!

File: UMIPs/umip-157.md (hunk 6 above)
Comment (nicholaspai) [reply UMIPs/umip-157.md:253]: I think this is implied by this line `Looking through all [valid fills](#valid-bundle-proposals) in history`, as `valid` can only refer to past events, right?

Comment (mrice32) [reply UMIPs/umip-157.md:253]: Maybe we should specify "in previously validated bundles"

**Review by mrice32 — DISMISSED** [review]
https://github.com/UMAprotocol/UMIPs/pull/550#discussion_r945972377

File: UMIPs/umip-157.md (hunk 6 above)
Comment (nicholaspai) [reply UMIPs/umip-157.md]: Done

**Review by mrice32 — APPROVED** [review]
LGTM!

**Pull request merged**